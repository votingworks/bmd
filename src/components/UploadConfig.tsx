import React from 'react'
import Dropzone from 'react-dropzone'
import styled from 'styled-components'

import electionSample from '../data/electionSample.json'
import isJSON from '../utils/isJSON'

import Button from './Button'
import Prose from './Prose'
import { Election } from '../config/types'
import Main, { MainChild } from './Main'
import MainNav from './MainNav'
import Text from './Text'

const Label = styled.label`
  display: flex;
  flex: 1;
  margin: auto;
  width: 100%;
`

const Footer = styled.footer`
  padding: 1rem;
`

interface Props {
  setElection: (election: Election) => void
}
interface State {
  errorMessage: string
  loaded: boolean
  loading: boolean
}

const initialState = {
  errorMessage: '',
  loaded: false,
  loading: false,
}

class UploadConfig extends React.Component<Props, State> {
  public state: State = initialState

  public setErrorMessage = (
    errorMessage: string = 'Only files that end in ".json" are accepted. Try again.'
  ) => {
    this.setState({
      ...initialState,
      errorMessage,
    })
  }

  public loadSampleElection = () => {
    this.props.setElection(electionSample as Election)
  }

  public onDrop = (acceptedFiles: File[]) => {
    if (acceptedFiles.length === 1) {
      acceptedFiles.forEach((file: File) => {
        if (file.name.endsWith('.json')) {
          const reader = new FileReader()
          reader.onload = () => {
            this.setState({
              loading: true,
            })
            const result = reader.result
            if (typeof result === 'string' && isJSON(result)) {
              window.setTimeout(() => {
                this.setState({
                  ...initialState,
                  loaded: true,
                })
                const election = JSON.parse(result)
                // TODO: Test that uploaded election.json file matches expected object shape. https://github.com/votingworks/bmd/issues/23
                this.props.setElection(election)
              }, 100)
            } else {
              this.setErrorMessage(
                'File content must be JSON text only. Try again.'
              )
            }
          }
          reader.readAsText(file)
        } else {
          this.setErrorMessage()
        }
      })
    } else {
      this.setErrorMessage()
    }
  }

  public render() {
    const { loading, loaded, errorMessage } = this.state
    if (process.env.NODE_ENV === 'test') {
      return (
        <>
          <Main noPadding>
            {loading || loaded ? (
              <p>{loaded ? 'File loaded' : 'Loading file…'} </p>
            ) : (
              <Dropzone
                accept="application/json"
                onDrop={this.onDrop}
                multiple={false}
              >
                {({ getRootProps, getInputProps, isDragActive }) => (
                  <Label
                    // TODO:: remove "as" after issue resolved:
                    // https://github.com/react-dropzone/react-dropzone/issues/182
                    as="div"
                    data-testid="dropzone"
                    htmlFor="election-file-upload"
                    {...getRootProps()}
                  >
                    <MainChild center padded>
                      <input
                        id="election-file-upload"
                        data-testid="file-input"
                        {...getInputProps()}
                      />
                      <Prose textCenter>
                        {isDragActive ? (
                          <p>Drop files here…</p>
                        ) : (
                          <React.Fragment>
                            <h1>Load Election Configuration File</h1>
                            <p>
                              Drag and drop <code>election.json</code> file
                              here, or click to browse for file.
                            </p>
                            {errorMessage && <Text error>{errorMessage}</Text>}
                          </React.Fragment>
                        )}
                      </Prose>
                    </MainChild>
                  </Label>
                )}
              </Dropzone>
            )}
          </Main>
          <Footer>
            <Prose textCenter>
              <Text center small>
                <a href="/data/election.json">
                  Download sample <code>election.json</code> file
                </a>{' '}
                to upload, or{' '}
                <Button onClick={this.loadSampleElection}>
                  Load Sample Election File
                </Button>
              </Text>
            </Prose>
          </Footer>
        </>
      )
    }
    return (
      <React.Fragment>
        <Main>
          <MainChild center>
            <Prose textCenter>
              <h1>Not Configured</h1>
              <p>Insert Election Clerk card.</p>
            </Prose>
          </MainChild>
        </Main>
        <MainNav />
      </React.Fragment>
    )
  }
}

export default UploadConfig
