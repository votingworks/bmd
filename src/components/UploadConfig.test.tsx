import { axe } from 'jest-axe'
import React from 'react'
import { fireEvent, render, waitForElement } from 'react-testing-library'

import goodElectionFile from '../../public/data/election.json'

import UploadConfig from './UploadConfig'

it(`allows file to be uploaded via file input`, async () => {
  const setElection = jest.fn()
  const { container, getByText, getByTestId } = render(
    <UploadConfig setElection={setElection} />
  )
  const fileInput = getByTestId('file-input')
  fireEvent.change(fileInput, {
    target: {
      files: [
        new File([JSON.stringify(goodElectionFile)], 'election.json', {
          type: 'application/json',
        }),
      ],
    },
  })
  await waitForElement(() => getByText('Loading file…'))
  await waitForElement(() => getByText('File loaded'))
  expect(setElection).toHaveBeenCalled()
  expect(container.firstChild).toMatchSnapshot()
})

it(`allows file to be uploaded via drag and drop`, async () => {
  const setElection = jest.fn()
  const { container, getByText, getByTestId } = render(
    <UploadConfig setElection={setElection} />
  )
  const dropzone = getByTestId('dropzone')
  fireEvent.dragEnter(dropzone)
  await waitForElement(() => getByText('Drop files here…'))
  fireEvent.drop(dropzone, {
    target: {
      files: [
        new File([JSON.stringify(goodElectionFile)], 'election.json', {
          type: 'application/json',
        }),
      ],
    },
  })
  await waitForElement(() => getByText('Loading file…'))
  await waitForElement(() => getByText('File loaded'))
  expect(setElection).toHaveBeenCalled()
  expect(container.firstChild).toMatchSnapshot()
})

it(`allows one-click config`, async () => {
  const setElection = jest.fn()
  const { container, getByText, getByTestId } = render(
    <UploadConfig setElection={setElection} />
  )
  fireEvent.click(getByText('Load Sample Election File'))
  expect(setElection).toHaveBeenCalled()
})

it(`doesn't allow files other than 'election.json'`, async () => {
  const setElection = jest.fn()
  const { getByText, getByTestId } = render(
    <UploadConfig setElection={setElection} />
  )
  const fileInput = getByTestId('file-input')
  fireEvent.change(fileInput, {
    target: {
      files: [
        new File([''], 'foo.json', {
          type: 'application/json',
        }),
      ],
    },
  })
  await waitForElement(() =>
    getByText('Only election.json file is accepted. Try again.')
  )
})

it(`doesn't allow upload of multiple files`, async () => {
  const setElection = jest.fn()
  const { getByText, getByTestId } = render(
    <UploadConfig setElection={setElection} />
  )
  const fileInput = getByTestId('file-input')
  fireEvent.change(fileInput, {
    target: {
      files: [
        new File([''], 'one.json', {
          type: 'application/json',
        }),
        new File([''], 'two.json', {
          type: 'application/json',
        }),
      ],
    },
  })
  await waitForElement(() =>
    getByText('Only election.json file is accepted. Try again.')
  )
})

it(`election config file must contain json content`, async () => {
  const setElection = jest.fn()
  const { getByText, getByTestId } = render(
    <UploadConfig setElection={setElection} />
  )
  const fileInput = getByTestId('file-input')
  fireEvent.change(fileInput, {
    target: {
      files: [
        new File(['asdf'], 'election.json', {
          type: 'application/json',
        }),
      ],
    },
  })
  await waitForElement(() =>
    getByText('File content must be JSON text only. Try again.')
  )
})

it(`UploadConfig is accessible`, async () => {
  const { container } = render(<UploadConfig setElection={jest.fn()} />)
  expect(await axe(container.innerHTML)).toHaveNoViolations()
})
