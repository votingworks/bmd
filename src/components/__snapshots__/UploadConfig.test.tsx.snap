// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`allows file to be uploaded via drag and drop 1`] = `
.c5 {
  box-sizing: border-box;
  cursor: pointer;
  background: lightgrey;
  border: none;
  border-radius: 0.25rem;
  padding: 0.4rem 0.7rem;
  color: black;
  line-height: 1;
  white-space: nowrap;
}

.c3 {
  line-height: 1.3;
  max-width: 66ch;
  text-align: center;
  margin: auto;
}

.c3 h1 {
  margin: 2rem 0 1rem;
}

.c3 h2 {
  margin: 1.5rem 0 1rem;
}

.c3 h3,
.c3 p {
  margin: 1rem 0;
}

.c3 h1 + h2 {
  margin-top: -0.75rem;
}

.c3 h1 + p,
.c3 h2 + p,
.c3 h3 + p {
  margin-top: -0.75rem;
}

.c3 > :first-child {
  margin-top: 0;
}

.c3 > :last-child {
  margin-bottom: 0;
}

.c0 {
  height: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c4 {
  text-align: center;
  font-size: 0.8rem;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  overflow: scroll;
}

.c2 {
  padding: 1rem;
}

@media print {
  .c1 {
    -webkit-box-pack: start;
    -webkit-justify-content: flex-start;
    -ms-flex-pack: start;
    justify-content: flex-start;
    padding: 0;
  }
}

<div
  class="c0"
>
  <main
    class="c1"
  >
    <p>
      File loaded
       
    </p>
  </main>
  <footer
    class="c2"
  >
    <div
      class="c3"
    >
      <p
        class="c4"
      >
        <a
          href="/data/election.json"
        >
          Download sample 
          <code>
            election.json
          </code>
           file
        </a>
         
        to upload, or
         
        <button
          class="c5"
          type="button"
        >
          Load Sample Election File
        </button>
      </p>
    </div>
  </footer>
</div>
`;

exports[`allows file to be uploaded via file input 1`] = `
.c5 {
  box-sizing: border-box;
  cursor: pointer;
  background: lightgrey;
  border: none;
  border-radius: 0.25rem;
  padding: 0.4rem 0.7rem;
  color: black;
  line-height: 1;
  white-space: nowrap;
}

.c3 {
  line-height: 1.3;
  max-width: 66ch;
  text-align: center;
  margin: auto;
}

.c3 h1 {
  margin: 2rem 0 1rem;
}

.c3 h2 {
  margin: 1.5rem 0 1rem;
}

.c3 h3,
.c3 p {
  margin: 1rem 0;
}

.c3 h1 + h2 {
  margin-top: -0.75rem;
}

.c3 h1 + p,
.c3 h2 + p,
.c3 h3 + p {
  margin-top: -0.75rem;
}

.c3 > :first-child {
  margin-top: 0;
}

.c3 > :last-child {
  margin-bottom: 0;
}

.c0 {
  height: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c4 {
  text-align: center;
  font-size: 0.8rem;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  overflow: scroll;
}

.c2 {
  padding: 1rem;
}

@media print {
  .c1 {
    -webkit-box-pack: start;
    -webkit-justify-content: flex-start;
    -ms-flex-pack: start;
    justify-content: flex-start;
    padding: 0;
  }
}

<div
  class="c0"
>
  <main
    class="c1"
  >
    <p>
      File loaded
       
    </p>
  </main>
  <footer
    class="c2"
  >
    <div
      class="c3"
    >
      <p
        class="c4"
      >
        <a
          href="/data/election.json"
        >
          Download sample 
          <code>
            election.json
          </code>
           file
        </a>
         
        to upload, or
         
        <button
          class="c5"
          type="button"
        >
          Load Sample Election File
        </button>
      </p>
    </div>
  </footer>
</div>
`;
